### ISSUE TITLE
app-compat: Printing issue and assertion failure with `ls` when running with `app-elfloader`

### ISSUE THREAD

**Issue by razvand** [unikraft/unikraft#857]
### Describe the bug

When using [`app-elfloader`](https://github.com/unikraft/app-elfloader) (via the `run_app.sh` script in [`run-app-elfloader`](https://github.com/unikraft/run-app-elfloader)) to run the [`ls` dynamic build](https://github.com/unikraft/dynamic-apps/tree/master/ls) a binary prefix is printed for each entry. Also, a dentry-related assertion appears.

The [`echo` dynamic build](https://github.com/unikraft/dynamic-apps/tree/master/echo) also bumps into the same dentry-related assertion.

### Steps to reproduce

You can easily just use the [`./do.sh`](https://github.com/unikraft-upb/scripts/tree/main/make-based/app-elfloader) script in the [`unikraft-upb/scripts` repository](https://github.com/unikraft-upb/scripts):

```console
$ ./do.sh setup
$ ./do.sh build
$ ./do.sh run_built ls
```

### Expected behavior

The `ls` command should properly print the directory entries. And no assertion should happen.

### Which architectures were you using or does this bug affect?

x86_64

### Which platforms were you using or does this bug affect?

kvm

### Relevant log output

```shell
openat(AT_FDCWD, ".", O_RDONLY|O_CLOEXEC|O_DIRECTORY|O_NONBLOCK) = fd:3
fstat(fd:3, <out>stat:{st_size=4096, st_mode=040775, ...}) = OK
getdents(0x3, 0x1000eda4a0, ...) = 0x578
getdents(0x3, 0x1000eda4a0, ...) = 0x0
close(fd:3) = OK
fstat(fd:1, <out>stat:{st_size=0, st_mode=020000, ...}) = OK
ioctl(0x1, 0x5401, ...) = 0x0
''$'\004''.'  ''$'\004''..'  ''$'\004''bin'  ''$'\004''lib'  ''$'\004''lib64'
write(fd:1, "''$'\004''.'  ''$'\004''"..., 78) = 78
close(fd:1) = OK
[    0.600017] CRIT: [libvfscore] <dentry.c @  216> Assertion failure: dp->d_refcnt > 0
```

**Comment by marcrittinghaus** [comment]
The dentry issue is fixed by #865. Unfortunately, I cannot reproduce the binary prefix that you see.

I used a Unikernel based on staging (53be06c) and merged the following PRs:
- #830
- #834
- #836
- #847
- #850
- #851
- #855
- #865

**Comment by andraprs** [comment]
I checked out the latest unikraft "staging" branch version [1] [2] together with the dynamically linked "ls" and "echo". The printing issue and the dentry assertion don't appear anymore.

I'll close this GitHub issue for now. If there is a new occurrence of this issue, please reopen it. Thank you.

[1] https://github.com/unikraft/unikraft/commits/staging
[2] https://github.com/unikraft/unikraft/commit/573c762f82cc4c4ddfdeb3c086fcf0dda5f56c7e
